src/main/java/io/jenkins/plugins/oidc_provider/Keys.java

```diff
@@ -132,4 +151,114 @@ static JSONObject key(IdTokenCredentials creds) {
+            .accumulate("crv", algorithm.curve)
+            .accumulate("x", x)
+            .accumulate("y", y);
+    }
+
+    public static JSONObject encodeRsaJsonWebKey(SupportedKeyAlgorithm algorithm, String keyId, RSAPublicKey key) {
+        Base64.Encoder encoder = Base64.getUrlEncoder().withoutPadding();
+        return new JSONObject().
+            accumulate("kid", keyId).
+            accumulate("kty", "RSA").
+            accumulate("alg", algorithm.name()).
+            accumulate("use", "sig").
+            accumulate("n", encoder.encodeToString(key.getModulus().toByteArray())).
+            accumulate("e", encoder.encodeToString(key.getPublicExponent().toByteArray()));
+    }
+
+    public enum AlgorithmType {
+        RSA, ELLIPTIC_CURVE
+    }
+
+    public enum SupportedKeyAlgorithm {
+        ES256(SignatureAlgorithm.ES256, "P-256", AlgorithmType.ELLIPTIC_CURVE),
+        ES384(SignatureAlgorithm.ES384, "P-384", AlgorithmType.ELLIPTIC_CURVE),
+        ES512(SignatureAlgorithm.ES512, "P-521", AlgorithmType.ELLIPTIC_CURVE),
+        RS256(SignatureAlgorithm.RS256, AlgorithmType.RSA),
+        RS384(SignatureAlgorithm.RS384, AlgorithmType.RSA),
+        RS512(SignatureAlgorithm.RS512, AlgorithmType.RSA);
+
+        private final SignatureAlgorithm algorithm;
+        private final String curve;
+        private final AlgorithmType type;
+
+        SupportedKeyAlgorithm(SignatureAlgorithm algorithm, AlgorithmType type) {
+            this(algorithm, null, type);
+        }
+
+        SupportedKeyAlgorithm(SignatureAlgorithm algorithm, String curve, AlgorithmType type) {
+            this.algorithm = algorithm;
+            this.curve = curve;
+            this.type = type;
+        }
+
+        public KeyPair generateKeyPair() {
+            return io.jsonwebtoken.security.Keys.keyPairFor(algorithm);
+        }
+
+        public AlgorithmType getType() {
+            return type;
+        }
+    }
+
+    public static class SecretKeyPair implements Serializable {
+        private static final long serialVersionUID = 2448941858110252020L;
+
+        /**
+         * Encrypted base64 encoding of a private key in {@link PKCS8EncodedKeySpec}
+         */
+        private final Secret privateKey;
+
+        /**
+         * Encrypted base64 encoding of a public key in {@link X509EncodedKeySpec}
+         */
+        private final Secret publicKey;
+        private final SupportedKeyAlgorithm algorithm;
+
+        private SecretKeyPair(SupportedKeyAlgorithm algorithm, byte[] privateKey, byte[] publicKey) {
+            this.privateKey = Secret.fromString(Base64.getEncoder().encodeToString(privateKey));
+            this.publicKey = Secret.fromString(Base64.getEncoder().encodeToString(publicKey));
+            this.algorithm = algorithm;
+        }
+
+        public KeyPair toKeyPair() throws Exception {
+            KeyFactory keyFactory;
+
+            switch (algorithm.type) {
+                case RSA:
```

Comment:
Since RSA can generate the public key from the private key, perhaps it would make sense for `SecretKeyPair` to be an `interface` with two implementations, only the EC version storing the public key?

Reply:
I've extracted a interface and there are now two distinct classes :slightly_smiling_face: 
This means only the EC version now stores the public key.